pageserver/src/billing_metrics.rs

```diff
@@ -0,0 +1,283 @@
+//!
+//! Periodically collect consumption metrics for all active tenants
+//! and push them to a HTTP endpoint.
+//! Cache metrics to send only the updated ones.
+//!
+
+use anyhow;
+use tracing::*;
+use utils::id::TimelineId;
+
+use crate::task_mgr;
+use crate::tenant_mgr;
+use pageserver_api::models::TenantState;
+use utils::id::TenantId;
+
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::fmt;
+use std::str::FromStr;
+use std::time::Duration;
+
+use chrono::{DateTime, Utc};
+use reqwest::Url;
+
+/// BillingMetric struct that defines the format for one metric entry
+/// i.e.
+///
+/// ```json
+/// {
+/// "metric": "remote_storage_size",
+/// "type": "absolute",
+/// "tenant_id": "5d07d9ce9237c4cd845ea7918c0afa7d",
+/// "timeline_id": "00000000000000000000000000000000",
+/// "time": ...,
+/// "value": 12345454,
+/// }
+/// ```
+#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub struct BillingMetric {
+    pub metric: BillingMetricKind,
+    pub metric_type: &'static str,
```

Comment:
This [serializes to](https://github.com/neondatabase/neon/actions/runs/3743454811/jobs/6355797651#step:4:7887) 
```
{
  "metric": "remote_storage_size",
  "metric_type": "absolute",
  "tenant_id": [
    0,
    69,
    128,
    207,
    205,
    193,
    78,
    45,
    241,
    206,
    228,
    159,
    51,
    170,
    110,
    230
  ],
  "timeline_id": null,
  "time": "2022-12-20T19:29:43.727764280Z",
  "value": 0
}
```

While [according to the spec](https://github.com/neondatabase/cloud/blob/a374c0e4c6b94116724137b186684453e785fe61/openapispec/billing/billing-v1.yaml#L133-L154) and to the example above (L28-L36) there should be fields:
```
metric
type (not metric_type)
tenant_id (a string, not an array of integers)
time
value
```
present. 

So there are inconsistencies now.

PS: The spec also requires `idempotency_key`, and while this could be softened on the control plane side, if we add changes here, could we please also add `idempotency_key` with current times + hostname? Because having this field is very much desired for the "exactly once" semantics, we would like to have, if we add retries, later on, any node of the event ingestion chain.

Reply:
UPD: I see that timeline_id is also serialized as an array of integers. OR as "null". Can we please send it 
1. either as string if present 
2. or not send it if not present? I understand 2 might be difficult on your side and will help by changing the spec on the consol plane side.

Reply:
Here is the PR
https://github.com/neondatabase/neon/pull/3215